### DISCUSSION TITLE
Concurrency in subscription

### DISCUSSION THREAD

**Discussion by weithegreat** [eProsima/Fast-DDS#3574] [Q&A]
I was wondering for the callbacks in data_listener of subscriber, will there be possible concurrency? When multiple publishers send through same topic concurrently, how are the messages handled? Are they handled in parallel/concurrently or sequentially?

**Comment by JLBuenoLopez** [comment]
@weigewansui,

If you have a DataReader receiving data from several DataWriters simultaneously, then the DDS specification only ensures that the data coming from the same DataWriter is sent sequentially. However, the data coming from the different DataReaders will come concurrently. You might consider adding to your data model a keyed field identifying the DataWriter and so you can access the data using [instances](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/topic/instances.html). In that case, each instance will receive the data sequentially. Also, there are several QoS configurations that affect how the communication is handled, so it will also depend on how you configure Fast DDS.

If you are referring to threading concurrency, you are not going to get the same callback for the same DataReader on different threads,  independently on the number of attached DataWriters.
